File: src/text_parser/text_parser.py
import abc


class TextParser(abc.ABC):
    def __init__(self):
        self.parameters = None
        self.content = ''
        self.reset()

    def set_content(self, content):
        self.content = content
        self.reset()

    def _filter(self, value, chain_filter=None):
        if chain_filter and value:
            value = chain_filter.filter(value)
        return value

    def reset(self):
        self.content_upper = str(self.content.upper())
        self.cursor = str(self.content)
        self.cursor_upper = str(self.content_upper)
        self.field = None
        self.field_upper = None

    def value(self, uppercase=False, chain_filter=None):
        filtered = self._filter(self.field, chain_filter)
        if uppercase and isinstance(filtered, str):
            return filtered.upper()
        return filtered

    @abc.abstractmethod
    def parse(self, tag_name):
        pass

File: src/text_parser.js
class TextParser {
  constructor() {
    this.parameters = null;
    this.content = '';
    this.reset();
  }

  set_content(content) {
    this.content = content;
    this.reset();
  }

  _filter(value, chain_filter) {
    if (chain_filter && value) {
      value = chain_filter.filter(value);
    }
    return value;
  }

  reset() {
    this.content_upper = this.content.toUpperCase();
    this.cursor = this.content;
    this.cursor_upper = this.content_upper;
    this.field = null;
    this.field_upper = null;
  }

  value(uppercase = false, chain_filter = null) {
    const filtered = this._filter(this.field, uppercase.chain_filter);
    if (uppercase.uppercase && typeof filtered === 'string') {
      return filtered.toUpperCase();
    }
    return filtered;
  }

  partition(str, separator) {
    const index = str.indexOf(separator);
    if (index === -1) {
      return [str, '', ''];
    }

    const partitioned = [
      str.substring(0, index),
      separator,
      str.substring(index + separator.length)
    ];

    return partitioned;
  }

  parse(tag_name) {
    // To be implemented in subclasses
  }
}

module.exports = TextParser;
